backstory/src/main/java/com/google/sps/servlets/ImageAnalysisServlet.java

```diff
@@ -0,0 +1,229 @@
+
+/** Servlet for managing image analysis with Vision API and blobstore for uploads */
+@WebServlet("/image-analysis")
+public class ImageAnalysisServlet extends HttpServlet {
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Query to find all analyzed image entities.
+    Query query = new Query("analyzed-image");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<AnalyzedImage> analyzedImages = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      String imageUrl = (String) entity.getProperty("imageUrl");
+      String labelsJsonArray = (String) ((Text) entity.getProperty("labelsJsonArray")).getValue();
+
+      analyzedImages.add(new AnalyzedImage(imageUrl, labelsJsonArray));
+    }
+
+    response.setContentType("application/json;");
+    Gson gson = new GsonBuilder().setPrettyPrinting().create();
+    String analyzedImagesJsonArray = gson.toJson(analyzedImages);
+    response.getWriter().println(analyzedImagesJsonArray);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Expecting a post request from Blobstore containing the data fields from the image-upload
+    // form. After having gone through Blobstore, the request will include the images uploaded. The
+    // form in the HTML will connect to the Blobstore URL, which encodes the images and then
+    // redirects the request to this Url. In this doPost, the images are analyzed with Vision API,
+    // and the analysis along with the image URL will be sent to datastore for permanent storage.
+
+    // Get the input from the form.
+    // Get the URL of the image that the user uploaded to Blobstore.
+    final String imageUrl = getUploadedFileUrl(request, "image-upload");
+
+    // Get the raw byte array representing the image from Blobstore
+    final byte[] bytes = getBlobBytes(request, "image-upload");
+
+    // Gets the full label information from the image byte array by calling Vision API.
+    List<EntityAnnotation> labels = detectLabelsBytes(bytes);
+    Gson gson = new Gson();
+    Text labelsJsonArray = new Text(gson.toJson(labels));
+
+    // Add the input to datastore
+    Entity analyzedImageEntity = new Entity("analyzed-image");
+    analyzedImageEntity.setProperty("imageUrl", imageUrl);
+    analyzedImageEntity.setProperty("labelsJsonArray", labelsJsonArray);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(analyzedImageEntity);
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/visionUploadPrototype/visionDemo.html#image-upload");
+  }
+
+  // Detects labels in the image specified by the image byte data by calling the Vision API.
+  private List<EntityAnnotation> detectLabelsBytes(byte[] bytes) throws IOException {
```

Comment:
detectLabelsFromImageBytes()?

(more grammatical and specifies what the bytes represent)